Сетка страниц и крупных блоков на флексах

VoblPi opened the pull request:
Пока только для главной страницы, т.к. многое не понимаю в построении сеток и не хочу плодить однотипные ошибки. Некоторые моменты написал в комментариях в коде.

---
:mortar_board: [Сетка страниц и крупных блоков на флексах](https://up.htmlacademy.ru/htmlcss/33/user/1872141/tasks/7)

:boom: https://htmlacademy-htmlcss.github.io/1872141-device-33/6/

Vital-Sh committed:
Флексы крупных блоков

andreysgra commented on index.html:
```diff
@@ -8,67 +8,93 @@
-        <li class="navigation_user_basket"><a class="link navigation_user" href="#">Корзина</a></li>
-      </ul>
+      <div class="header_user_container">
+        <ul class="header_user_menu">
+          <li class="header_search">
+            <form class="header_search_container" action="https://echo.htmlacademy.ru/">
+              <input class="search_input" type="search" id="search_input" placeholder="Поиск по сайту">
+              <button class="button search_button" type="submit">Найти</button>
+            </form>
+          </li>
```
Список в данном случае не нужен, так как здесь кроме формы нет и не предвидится схожих по смыслу элементов.

andreysgra commented on index.html:
```diff
@@ -8,67 +8,93 @@
         </div>
         <ol class="slider_properties">
           <li class="slider_item">
+            <!--
+              Стрелки переключения слайдера должны быть частью слайда,
+              но как их впихнуть, сохранив размер картинки не разобрался.
+              Слайд накладывать фоном? Так же не понял,
+              как сделать отступ до кнопки/стрелки "Предыдущее фото" 30px, а между кнопкой/стрелкой
+              "Следующее фото" отступ 40 px. Прописывать margin-left/margin-right для кнопок?
+            -->
```
Я считал, что стрелки нужны для переключения картинок в слайде, так как картинок товара может быть больше одной. Но в процессе обсуждения этого момента среди наставников Николай Шабалин сказал, что эти стрелки служат для переключения слайдов, так же, как и круглые кнопки.

Лично для меня, как пользователя, такое решение кажется странным. Стрелки расположены по краям фото и это наводит на мысль, что они переключают фотографии.

Сейчас расположением стрелок заниматься не нужно. Всё таки задание про крупные сетки. А в дальнейшем блок со стрелками можно будет позиционировать абсолютно поверх всего слайдера. Для этих стрелок потребуется собственная обёртка `div` которую и будем позиционировать абсолютно.

andreysgra commented on index.html:
```diff
@@ -8,67 +8,93 @@
               </button>
             </div>
-            <img class="slider_image_quadrocopter" src="images/slider/slider_1.png" width="560" height="560" alt="Квадрокоптер с лазером, замаскированным под камеру." >
+            <!--
+              Я не понимаю как номер слайдера впихать в разметку, т.к. он портит у меня все флексы,
+               а разбор вёрстки слайдера будет позже.
+
```
Оставьте пока всё как есть. По сути это декоративные элементы и к ним вернёмся позже.

andreysgra commented on index.html:
```diff
@@ -195,8 +226,20 @@ <h4 class="contacts_subtitle">Время работы</h4>
+          -->
+          <p>Еженедельный дайджест с обзорами новинок, длительными тест-драйвами, распаковками, а также акции, спецпредложения и вот это вот всё.</p>
+        </div>
+        <!--
+          Тут display: flex надо применять для класса mailing-list и форма для e-mail так же является частью
+          сетки, но построение таких сеток рассматривается позже. Поэтому пока в сетку из флексов взято только
+          2 элемента. Поэтой же причине не выполнил центровку формы. Или в равках урока надо задать ей
+           отдельно свойство display: flex?
+        -->
```
`mailing-list` нужно вписать в общую сетку, то есть создать из него контейнер контента. Потому что на разрешении экрана > 1440 сетка в этой области начинает разваливаться.
![image](https://user-images.githubusercontent.com/17790771/156874996-fc647136-dbd9-4cfc-9afa-5f9abd8a86d2.png)

andreysgra commented on styles/styles.css:
```diff
@@ -48,14 +56,41 @@ body {
+  margin-top: 40px;
+}
+
+.logo {
+  margin-left: 60px;
+}
```
Если мы внимательно посмотрим на расположение элементов шапки, то увидим, что отступы по краям составляют по 60px:
![image](https://user-images.githubusercontent.com/17790771/156875143-6d8cee36-c3b3-4c11-a160-1c68becbe6ea.png)
Следовательно можно добавить шапке `header_page` свойство `padding: 0 60px`. Но тут нужно учитывать важный нюанс: при стандартном расчёте размеров блока (значение свойства `box-sizing` равно `content-box`) паддинги увеличивают ширину блока и соответственно она станет равной 1160 + 60 * 2, что превысит макетную ширину.
Есть 2 варианта решения вопроса:
1) изменить значение `width` - 1160 - 60 * 2;
2) изменить алгоритма расчёта ширины и высоты элемента, то есть задать `box-sizing: border-box`.

Оба варианта равнозначны и какой из них использовать, решать вам.

andreysgra commented on styles/styles.css:
```diff
@@ -48,14 +56,41 @@ body {
+  margin-left: 60px;
+}
+
+.header_user_container {
+  display: flex;
+  width: 1040px;
```
Если сделаете паддинги у родительского блока, как я описывал выше, то блоку `header_user_container` задавать `width` не придётся и `margin: 0 auto` будет без надобности.

andreysgra commented on styles/styles.css:
```diff
@@ -64,13 +99,16 @@ body {
   font-weight: inherit;
   line-height: inherit;
   text-transform: none;
+  flex-grow: 1;
```
В общем то здесь необходимости во `flex-grow: 1` нет. Дело в том, что выше, в `header_user_container` вы задаёте распределение свободного пространства во флекс-контейнере с помощью `justify-content: space-between`. От этого стоит отказаться, а флекс-элементы отталкивать на нужно расстояние с помощью `margin`.

andreysgra commented on styles/styles.css:
```diff
@@ -80,15 +118,37 @@ body {
 }
 
-.navigation_menu_list {
+.navigation_menu {
   text-transform: uppercase;
+  display: flex;
+  width: 1040px;
+  justify-content: space-between;
+  margin: 0 auto;
```
В этом меню `width: 1040px` и `margin: 0 auto` тоже будут без надобности, если зададите паддинги в шапке `header_page`.

andreysgra commented on styles/styles.css:
```diff
@@ -99,6 +159,13 @@ body {
 
+.slider_item {
+  display: flex;
+  align-items: flex-end;
```
`align-items: flex-end` в `slider_item` можно и не делать, а сместить `slider_item_description_container` вниз с помощью `margin-top`. У вас здесь не ошибка, поэтому я просто написал возможный вариант, а какое решение лучше - нужно смотреть уже в дальнейшем, при построении микросеток.

andreysgra commented on styles/styles.css:
```diff
@@ -125,31 +192,91 @@ body {
+}
+
+.about-shop h3 {
+  font-family: 'raleway';
```
Не указано типовое семейство для шрифта. [CSS-06](https://up.htmlacademy.ru/htmlcss/33/criteries#CSS-06)

andreysgra commented on styles/styles.css:
```diff
@@ -170,11 +297,25 @@ body {
   font-style: normal;
 }
 
+.mailing-list_container {
+  display: flex;
+  width: 1440px;
```
У этого блока, как и у других крупных сеточных, ширина 1160px.

Vital-Sh committed:
Исправил ошибки в ДЗ №4

Если я правильно понял и определил крупные блоки, то остальные элементы будут верстаться в следующих частях обучения.

Vital-Sh committed:
Исправление ошибок в ДЗ№4

Исправил ошибки главной страницы и страницы каталога.

Vital-Sh committed:
Исправил размеры крупных блоков

Добавил фон на всю страницу у секции рассылки.

andreysgra commented on catalog.html:
```diff
@@ -40,177 +42,186 @@ <h1>Моноподы для селфи</h1>
+              <ul class="pagination_list">
+                <li class="pagination_item">
+                  <a class="pagination_link pagination_prev pagination_disabled">
+                    <span>Назад</span>
```
В принципе здесь нет никакой необходимости оборачивать текст ссылки в `<span>`. Просто лишний HTML элемент.

andreysgra commented on catalog.html:
```diff
@@ -40,177 +42,186 @@ <h1>Моноподы для селфи</h1>
+                </li>
+                <li class="pagination_item">
+                  <a class="pagination_link pagination_next" href="#">
+                    <span>Вперёд</span>
```
Здесь аналогично предыдущему комментарию.

andreysgra commented on styles/styles.css:
```diff
@@ -170,32 +321,135 @@ body {
+  line-height: 20px;
+}
+
+.main_inner {
+  font-family: "rubik", sans-serif;
+  font-size: 16px;
+  line-height: 19px;
+  color: #000000;
+  background-color: #ffffff;
+  flex-grow: 1;
+  width: 100%;
+  display: flex;
+  flex-flow: column;
+}
```
По большому счёту `main_inner` ничем не отличается от `main_page`, а свойства для текста и фон в общем-то уже заданы в `body`.

andreysgra commented on styles/styles.css:
```diff
@@ -205,22 +459,62 @@ body {
   line-height: 50px;
 }
 
+.bread-crumbs_list {
+  display: flex;
+  width: 1040px;
+  padding: 0;
+}
+.bread-crumbs_list li:not(first-child) {
+  margin-left: 36px;
+}
```
Советую присмотреться к свойству `column-gap` (его нужно применить к флекс-контейнеру `bread-crumbs_list`). Тогда не нужно будет задавать расстояние между флекс-элементами с помощью `margin`.

andreysgra commented on styles/styles.css:
```diff
@@ -205,22 +459,62 @@ body {
 }
 
+.product_background_container {
+  width: 100%;
```
Зададите вы `width: 100%` блоку `product_background_container` или не зададите, ничего не изменится. В стандартной блочной модели значение `width` по умолчанию равно `auto`. Это означает, что блок занимает всю доступную ему ширину, а так как `product_background_container` ничем не ограничен, то растянется на всю ширину экрана.

Также в стандартной блочной модели есть нюанс со свойством `width`. Это не ширина блока, как кажется на первый взгляд, это ширина контента. Если при `width: 100%` добавить блоку паддинги по краям, то блок выйдет за пределы родительского блока, так как его ширина станет равной `100% + padding-left + padding-right`.

andreysgra commented on styles/styles.css:
```diff
@@ -205,22 +459,62 @@ body {
+  width: 1040px;
+  margin: auto;
+}
+ /* Мне кажется, что в макете ошибка, т.к. размер блока для элемента 'фильтр:' указана 273px,
+ а для общего флексбокса (в который входят 'Стоимость', 'Цвет', 'Bluetooth') 260px. */
+.filter {
+  width: 260px;
```
Хм-м, а почему бы не 320px?
![image](https://user-images.githubusercontent.com/17790771/157067065-98aaaa2e-ab47-42fb-9285-c5b72bd5ae52.png)
Ширина соседнего блока с карточками товара 840px и всё в совокупности с блокам фильтра составляет 1160px, как раз ширину контентной области.

andreysgra commented on styles/styles.css:
```diff
@@ -239,6 +533,20 @@ body {
   background-color: #ffffff;
 }
 
+.product_page_monopod ul {
+  display: flex;
+  width: 840px;
+  justify-content: space-evenly;
```
В дальнейшем, раскладку карточек товаров лучше сделать на гридах (в соответствующем задании).

andreysgra commented on styles/styles.css:
```diff
@@ -239,6 +533,20 @@ body {
+  padding: 0;
+}
+
+.product_page_monopod ul:last-child {
```
Лучше присвоить данному списку класс.

Vital-Sh committed:
Исправил замечания

Вёрстка блоков карточек товара и пагинации будет выполнена в дальнейшем на гридах. Поэтому убрал ранее написанный код для этих блоков.
